src/v/config/property_constraints.cc

```diff
@@ -0,0 +1,462 @@
+from_string_view<constraint_type>(std::string_view sv) {
+    return string_switch<std::optional<constraint_type>>(sv)
+      .match("none", constraint_type::none)
+      .match("restrict", constraint_type::_restrict)
+      .match("clamp", constraint_type::clamp)
+      .default_match(constraint_type::none);
+}
+
+// This is where all the type conversions happen
+std::shared_ptr<constraint_validator_t> constraint_validator_map(
+  ss::sstring& name,
+  std::optional<ss::sstring> min_str_opt,
+  std::optional<ss::sstring> max_str_opt,
+  constraint_enabled_t enabled) {
+    auto& cfg = config::shard_local_cfg();
+
+    if (name == "log_cleanup_policy") {
+        return std::make_shared<
+          constraint_validator_enabled<model::cleanup_policy_bitflags>>(
+          std::move(enabled),
+          cfg.log_cleanup_policy.bind(),
+          [](const cluster::topic_configuration& topic_cfg) {
+              return topic_cfg.properties.cleanup_policy_bitflags
+                       ? *topic_cfg.properties.cleanup_policy_bitflags
+                       : model::cleanup_policy_bitflags::none;
+          },
+          [](
+            cluster::topic_configuration& topic_cfg,
+            model::cleanup_policy_bitflags& topic_val) {
+              topic_cfg.properties.cleanup_policy_bitflags = topic_val;
+          });
+    } else if (name == "log_retention_ms") {
+        std::optional<std::chrono::milliseconds> min_opt{std::nullopt},
+          max_opt{std::nullopt};
+        if (min_str_opt) {
+            min_opt = std::chrono::milliseconds{std::stol(*min_str_opt)};
+        }
+        if (max_str_opt) {
+            max_opt = std::chrono::milliseconds{std::stol(*max_str_opt)};
+        }
+        auto def_val = *cfg.log_retention_ms.default_value();
+        return std::make_shared<
+          constraint_validator_range<std::chrono::milliseconds>>(
+          std::move(min_opt),
+          std::move(max_opt),
+          [def_val{std::move(def_val)}](
+            const cluster::topic_configuration& topic_cfg) {
+              return topic_cfg.properties.retention_duration
+                         .has_optional_value()
+                       ? topic_cfg.properties.retention_duration.value()
+                       : def_val;
+          },
+          [](
+            cluster::topic_configuration& topic_cfg,
+            std::chrono::milliseconds& topic_val) {
+              topic_cfg.properties.retention_duration
+                = tristate<std::chrono::milliseconds>(topic_val);
+          });
+    } else if (name == "log_segment_ms") {
+        std::optional<std::chrono::milliseconds> min_opt{std::nullopt},
+          max_opt{std::nullopt};
+        if (min_str_opt) {
+            min_opt = std::chrono::milliseconds{std::stol(*min_str_opt)};
+        }
+        if (max_str_opt) {
+            max_opt = std::chrono::milliseconds{std::stol(*max_str_opt)};
+        }
+        auto def_val = *cfg.log_segment_ms.default_value();
```

Comment:
> Unconditional `optional::operator*`. The config does have a default value (I assume that's unlikely to change), but perhaps a runtime assertion would be appropriate in these cases?

Good idea. I'll change.

> Related random thought: in general, is it meaningful to apply a constraint to a default value?

I've gone back-and-forth on what `value_extractor` should return in the **rare** chance that a topic configuration has an undefined property, the config's default value or some hard-coded value. This scenario seems to happen in unit tests and not regular execution. The issue is related to semantics, For example, returning 0 milliseconds for log_retention_ms would likely fail any minimum but is it OK to clamp retention to 0 milliseconds? Let's discuss in standup at some point but I see arguments for both config's default value or a hard-coded one:
1. config's default value - our config properties have default values which are formed from the type's default constructor [here](https://github.com/redpanda-data/redpanda/blob/c30a6c5bb342b8c659b6964c930f857c0ffc0f96/src/v/config/property.h#L87). So, regardless of the constraint type, the system would use default values that already exist in Redpanda.
2. hard-code - This is tricky because do we optimize for _restrict_ or _clamp_ constraint  type? In the context of retention, my concern would be that Redpanda deletes a log sooner/later than the user expected. Users at least have an idea with option 1. 

I went with option 1 but I can be convinced otherwise.

Reply:
(1) sounds reasonable to me, though from the constructor you linked it looks like that default could be `nullopt` (for other `std::optional` properties, not `log_segment_ms`), which implies some bespoke logic to handle cases where the default is null vs a meaningful value. 

Anyway yeah, seems like there are some subtle edges here. Happy to chat through it anytime.

Reply:
Discussed in standup. Constraints are accepted at runtime if existing topics already satisfy the constraint. Topics properties are assigned default values when users do **not** specify them. Therefore, returning the default value here is not a problem.